mkdocs/docs/expression-dsl.md

```diff
@@ -0,0 +1,261 @@
+
+# age is not 18, 19, oer 20
+age_not_in_set = NotIn("age", [18, 19, 20])
+```
+
+#### Null Predicates
+
+```python
+from pyiceberg.expressions import IsNull, NotNull
+
+# Is null
+name_is_null = IsNull("name")
+
+# Is not null
+name_is_not_null = NotNull("name")
+```
+
+#### String Predicates
+
+```python
+from pyiceberg.expressions import StartsWith, NotStartsWith
+
+# TRUE for 'Johnathan', FALSE for 'Johan'
+name_starts_with = StartsWith("name", "John")
+
+# FALSE for 'Johnathan', TRUE for 'Johan'
+name_not_starts_with = NotStartsWith("name", "John")
+```
+
+### Logical Operators
+
+You can combine predicates using logical operators:
+
+```python
+from pyiceberg.expressions import And, Or, Not
+
+# TRUE for 25, FALSE for 67 and 15
+age_between = And(
+    GreaterThanOrEqual("age", 18),
+    LessThanOrEqual("age", 65)
+)
+
+# FALSE for 25, TRUE for 67 and 15
+age_outside = Or(
+    LessThan("age", 18),
+    GreaterThan("age", 65)
+)
+
+# NOT operator
+not_adult = Not(GreaterThanOrEqual("age", 18))
+```
+
+## Advanced Usage
+
+### Complex Expressions
+
+You can build complex expressions by combining multiple predicates and operators:
+
+```python
+from pyiceberg.expressions import And, Or, Not, EqualTo, GreaterThan, LessThan, In
+
+# (age >= 18 AND age <= 65) AND (status = 'active' OR status = 'pending')
+complex_filter = And(
+    And(
+        GreaterThanOrEqual("age", 18),
+        LessThanOrEqual("age", 65)
+    ),
+    Or(
+        EqualTo("status", "active"),
+        EqualTo("status", "pending")
+    )
+)
+
+# NOT (age < 18 OR age > 65)
+age_in_range = Not(
+    Or(
+        LessThan("age", 18),
+        GreaterThan("age", 65)
+    )
+)
+```
+
+### Type Safety
+
+The expression DSL provides type safety through Python's type system. When you create expressions, the types are checked at runtime:
+
+```python
+from pyiceberg.expressions import EqualTo
+
+# This will work
+age_equals_18 = EqualTo("age", 18)
+
+# This will raise a TypeError if the field type doesn't match
+age_equals_18 = EqualTo("age", "18")  # Will fail if age is an integer field
+```
+
+## Best Practices
+
+1. **Use Type Hints**: Always use type hints when working with expressions to catch type-related errors early.
+
+2. **Break Down Complex Expressions**: For complex expressions, break them down into smaller, more manageable parts:
+
+```python
+# Instead of this:
+complex_filter = And(
+    And(
+        GreaterThanOrEqual("age", 18),
+        LessThanOrEqual("age", 65)
+    ),
+    Or(
+        EqualTo("status", "active"),
+        EqualTo("status", "pending")
+    )
+)
+
+# Do this:
+age_range = And(
+    GreaterThanOrEqual("age", 18),
+    LessThanOrEqual("age", 65)
+)
+
+status_filter = Or(
+    EqualTo("status", "active"),
+    EqualTo("status", "pending")
+)
+
+complex_filter = And(age_range, status_filter)
+```
+
+## Common Pitfalls
+
+1. **Type Mismatches**: Always ensure that the types of your literals match the field types in your schema.
```

Comment:
PyIceberg should automatically cast the literal to the right type when it binds to the schema. Did you run into anything specific?